tachyon/math/polynomials/univariate/univariate_polynomial.h

```diff
@@ -122,10 +124,33 @@ class UnivariatePolynomial final
   template <typename Container>
   constexpr static Field EvaluateVanishingPolyByRoots(const Container& roots,
                                                       const Field& point) {
-    return std::accumulate(roots.begin(), roots.end(), Field::One(),
-                           [point](Field& acc, const Field& root) {
-                             return acc *= (point - root);
-                           });
+#if defined(TACHYON_HAS_OPENMP)
+    size_t thread_nums = static_cast<size_t>(omp_get_max_threads());
+#else
+    size_t thread_nums = 1;
+#endif
+    size_t n = roots.size();
+    if (n * 2 < thread_nums) {
+      return std::accumulate(roots.begin(), roots.end(), Field::One(),
+                             [point](Field& acc, const Field& root) {
+                               return acc *= (point - root);
+                             });
+    } else {
+      size_t chunk_size = (n + thread_nums - 1) / thread_nums;
+      std::vector<Field> parts(thread_nums, Field::One());
+      OPENMP_PARALLEL_FOR(size_t i = 0; i < thread_nums; ++i) {
+        size_t start_idx = i * chunk_size;
+        size_t end_idx = std::min(start_idx + chunk_size, n);
+        Field product = Field::One();
+        for (size_t j = start_idx; j < end_idx; ++j) {
+          product *= (point - roots[j]);
+        }
+        parts[i] = product;
+      }
```

Comment:
Please use `base::Parallelize()` and extract "for loop" to `EvaluateVanishingPolyByRootsSerial`